### DISCUSSION TITLE
Repository with SHACL as well as Lucene

### DISCUSSION THREAD

**Discussion by volkerjaenisch** [eclipse-rdf4j/rdf4j#3321] [Q&A]
Dear Community!

- Is it possible to configure a repository to have  SHACL as well as Lucene support? If so how to configure it?
- Are then both features available via the web-API?

Cheers,
Volker

**Comment by barthanssens** [comment]
IMHO something like this should work 

```
Sail baseSail = new NativeStore();

LuceneSail lucenesail = new LuceneSail();
lucenesail.setParameter(LuceneSail.LUCENE_RAMDIR_KEY, "true");
lucenesail.setBaseSail(baseSail);

ShaclSail shaclSail = new ShaclSail(lucenesail);

Repository repo = new SailRepository(shaclSail);
```

**Reply by barthanssens** [reply]
Oh, and example of retrieving validation report from remote server can be found in the Junit tests:

https://github.com/eclipse/rdf4j/blob/3.7.2/tools/server/src/test/java/org/eclipse/rdf4j/http/server/ShaclValidationReportIT.java

**Comment by volkerjaenisch** [comment]
Thank you @barthanssens !

I am currently writing a Python API for RDF4J and therefore I am strictly bound to the WEB-API.

This is the TTL to create a lucene repository via the WEB-API.

```
[] a rep:Repository ;
   rep:repositoryID "{repo_id}" ;
   rdfs:label "{repo_label}" ;
   rep:repositoryImpl [
      rep:repositoryType "openrdf:SailRepository" ;
      sr:sailImpl [
        sail:sailType "openrdf:LuceneSail";
        sail-luc:indexDir "index/" ;
        sail:delegate [
            sail:sailType "openrdf:MemoryStore" ;
            sail:iterationCacheSyncThreshold "{iterationCacheSyncThreshold}";
            ms:persist {persist} ;
            ms:syncDelay {syncDelay};
            sb:evaluationStrategyFactory "{evaluationStrategyFactory}"
        ]
    ]
].
```

And this is the WEB-API configuration for a SHACL repository.

```
[] a rep:Repository ;
   rep:repositoryID "{repo_id}" ;
   rdfs:label "{repo_label}" ;
   rep:repositoryImpl [
      rep:repositoryType "openrdf:SailRepository" ;
      sr:sailImpl [
         sail:sailType "rdf4j:ShaclSail" ;
         sail:delegate [
            sail:sailType "openrdf:MemoryStore" ;
            sail:iterationCacheSyncThreshold "{iterationCacheSyncThreshold}";
            ms:persist {persist} ;
            ms:syncDelay {syncDelay};
            sb:evaluationStrategyFactory "{evaluationStrategyFactory}"
         ]
      ]
   ].

```

Can I stack the sails like in your Java code in TTL? 

Cheers,
Volker

**Reply by barthanssens** [reply]
I haven't tested it yet (I'm always using the java code), but stacking should indeed work via the REST API.
If not it should be considered a bug IMHO :-)
